lib/classes/tooltip.js

```diff
@@ -0,0 +1,632 @@
+const CLASS_PREFIX = 'bs-tooltip';
+const BSCLS_PREFIX_REGEX = new RegExp(`(^|\\s)${CLASS_PREFIX}\\S+`, 'g');
+
+const TRANSITION_DURATION = 150;
+
+// Modal $root event
+const MODAL_CLOSE_EVENT = 'hidden::modal';
+const MODAL_CLASS = '.modal';
+
+const AttachmentMap = {
+    AUTO: 'auto',
+    TOP: 'top',
+    RIGHT: 'right',
+    BOTTOM: 'bottom',
+    LEFT: 'left'
+};
+
+const HoverState = {
+    SHOW : 'show',
+    OUT  : 'out'
+};
+
+const ClassName = {
+    FADE : 'fade',
+    SHOW : 'show'
+};
+
+const Selector = {
+    TOOLTIP: '.tooltip',
+    TOOLTIP_INNER: '.tooltip-inner',
+    ARROW: '.arrow'
+};
+
+const Trigger = {
+    HOVER: 'hover',
+    FOCUS: 'focus',
+    CLICK: 'click',
+    MANUAL: 'manual'
+};
+
+const Defaults = {
+    animation : true,
+    template: '<div class="tooltip" role="tooltip">' +
+              '<div class="arrow"></div>' +
+              '<div class="tooltip-inner"></div>' +
+              '</div>',
+    trigger: 'hover focus',
+    title: '',
+    delay: 0,
+    html: false,
+    placement: 'top',
+    offset: 0,
+    container: false,
+    fallbackPlacement : 'flip'
+};
+
+// Transition Event names
+const TransitionEndEvents = {
+    WebkitTransition: 'webkitTransitionEnd',
+    MozTransition: 'transitionend',
+    OTransition: 'otransitionend oTransitionEnd',
+    transition: 'transitionend'
+};
+
+// Client Side Tip ID counter for aria-describedby attribute
+// Could use Alex's uid generator util
+// Each tooltip requires a unique client side ID
+let NEXTID = 1;
+const generateId = name => `__BV_${name}_${NEXTID++}__`;
+
+/*
+ * Polyfill for Element.closest() for IE :(
+ * https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#Polyfill
+ */
+if (inBrowser && window.Element && !Element.prototype.closest) {
+    Element.prototype.closest = function (s) {
+        const matches = (this.document || this.ownerDocument).querySelectorAll(s);
+        let el = this;
+        let i;
+        do {
+            i = matches.length;
+            // eslint-disable-next-line no-empty
+            while (--i >= 0 && matches.item(i) !== el) {
+            }
+        } while ((i < 0) && (el = el.parentElement));
+        return el;
+    };
+}
+
+/*
+ * ToolTip Class definition
+ */
+class ToolTip {
+
+    // Main constructor
+    constructor(element, config, $root) {
+        // New tooltip object
+        this.$fadeTimeout = null;
+        this.$hoverTimeout = null;
+        this.$hoverState = '';
+        this.$activeTrigger = {};
+        this.$popper = null;
+        this.$element = element;
+        this.$tip = null;
+        this.$root = $root || null;
+        this.updateConfig(config);
+    };
+
+    static get Default() {
+        return Defaults;
+    };
+
+    static get NAME() {
+        return NAME;
+    }
+
+
+    // Update config
+    updateConfig(config) {
+        // Merge config into defaults
+        let updatedConfig = assign({}, ToolTip.constructor.Default, config);
```

Comment:
we have to use `this..constructor.Default` rather than `ToolTip.constructor.Default`, otherwise Popover will look like a tool tip.  so `this` is used in order to ensure that popover, when it overrides defaults gets the right template